### PR TITLE
feat: validation for property changes for i18n models

### PR DESCRIPTION
**Pull request by GDamyanov**
Validation for property changes for i18n models 
#1458

### CODE REVIEW

File: packages/preview-middleware-client/src/cpe/changes/validator.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import ResourceBundle from 'sap/base/i18n/ResourceBundle';
+
+/**
+ * Function to validate if a given value is a valid binding model.
+ *
+ * @param value value to be checked.
+ */
+export function vaildateBindingModel(value: string): void {
+    const bindingValue = value.replace(/{|}/gi, '');
```
Comment (zdravko-georgiev) [inline packages/preview-middleware-client/src/cpe/changes/validator.ts:9]: Plese change as proposed in here https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=1459&id=SAP_open-ux-tools&open=AYvH0V2Hd2OFDSMPrS8P&tab=why to fix the code smell

File: packages/preview-middleware-client/src/cpe/changes/validator.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    const bindingParts = bindingValue.trim().split('>').filter(el => el);
```
Comment (zdravko-georgiev) [inline packages/preview-middleware-client/src/cpe/changes/validator.ts:10]: Are you sure about this one? What is you filter condition?

File: packages/preview-middleware-client/src/cpe/changes/validator.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import ResourceBundle from 'sap/base/i18n/ResourceBundle';
+
+/**
+ * Function to validate if a given value is a valid binding model.
+ *
+ * @param value value to be checked.
+ */
+export function vaildateBindingModel(value: string): void {
```
Comment (zdravko-georgiev) [inline packages/preview-middleware-client/src/cpe/changes/validator.ts:8]: you have a typo 😉 It should be `validate`

File: packages/preview-middleware-client/src/cpe/changes/validator.ts (hunk 2 above)
Comment (GDamyanov) [reply packages/preview-middleware-client/src/cpe/changes/validator.ts]: When you split and the value ends with the split char the return array contains one additional empty string element.

File: packages/preview-middleware-client/src/cpe/changes/validator.ts (hunk 3 above)
Comment (GDamyanov) [reply packages/preview-middleware-client/src/cpe/changes/validator.ts]: Fixed

File: packages/preview-middleware-client/src/cpe/changes/validator.ts (hunk 1 above)
Comment (GDamyanov) [reply packages/preview-middleware-client/src/cpe/changes/validator.ts]: Fixed

File: packages/preview-middleware-client/src/cpe/changes/flex-change.ts [hunk 4]
```diff
@@ -2,15 +2,16 @@ import type FlexCommand from 'sap/ui/rta/command/FlexCommand';
...
  * Function to check a give value is a binding expression.
  *
- * @param value value to be checked.
+ * @param change changed property of a control
  * @returns boolean
  */
-function isBindingExpression(value: string): boolean {
-    return value.includes('{') && value.includes('}');
+function isBindingExpression(change: PropertyChange): boolean {
+    return typeof change.value === 'string' && change.value.includes('{') && change.value.includes('}');
```
Comment (nikmace) [inline packages/preview-middleware-client/src/cpe/changes/flex-change.ts:14]: Maybe we can extract change.value here and add optional chaining (maybe the value can be null/undefined | change.value is duplicated in three places). 

Also we can include a check that '{' occurs before '}' like this `value.indexOf('{') < value.indexOf('}')`

File: packages/preview-middleware-client/src/cpe/changes/validator.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Function to validate if a given value is a valid binding model.
+ *
+ * @param value value to be checked.
+ */
+export function validateBindingModel(value: string): void {
+    const bindingValue = value.replace(/[{}]/gi, '');
+    const bindingParts = bindingValue
+        .trim()
+        .split('>')
```
Comment (nikmace) [inline packages/preview-middleware-client/src/cpe/changes/validator.ts:12]: The trim method could be put after replace like this:

```typescript
const bindingValue = value.replace(/[{}]/gi, '').trim();
const bindingParts = bindingValue.split('>').filter(el => el);
```

File: packages/preview-middleware-client/src/cpe/changes/validator.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    if (!bindingParts.length) {
+        throw new SyntaxError('Invalid binding string.');
+    }
+
+    if (bindingParts[0].trim() === 'i18n') {
+        if (bindingParts.length === 2) {
+            const resourceBundle = ResourceBundle.create({
+                url: '../i18n/i18n.properties'
+            }) as ResourceBundle;
+            if (!resourceBundle.hasText(bindingParts[1].trim())) {
```
Comment (nikmace) [inline packages/preview-middleware-client/src/cpe/changes/validator.ts:24]: This `bindingParts[1].trim()` could be extracted in its own variable like **resourceKey**.

**Review by nikmace — CHANGES_REQUESTED** [review]

File: packages/preview-middleware-client/src/cpe/changes/flex-change.ts (hunk 4 above)
Comment (GDamyanov) [reply packages/preview-middleware-client/src/cpe/changes/flex-change.ts]: Done

File: packages/preview-middleware-client/src/cpe/changes/validator.ts (hunk 6 above)
Comment (GDamyanov) [reply packages/preview-middleware-client/src/cpe/changes/validator.ts]: Done

**Review by nikmace — APPROVED** [review]
The code looks good
Review comments addressed
Coverage is good
Tested manually

File: packages/preview-middleware-client/src/cpe/changes/validator.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+            if (!resourceBundle.hasText(resourceKey)) {
+                throw new SyntaxError(
+                    'Invalid key in the binding string. Supported value pattern is {i18n>YOUR_KEY}. Check if the key already exists in i18n.properties. If not, add the key in the i18n.properties file and reload the editor for the new key to take effect.'
+                );
+            }
+        } else {
+            throw new SyntaxError('Invalid binding string. Supported value pattern is {i18n>YOUR_KEY}');
+        }
+    }
+}
```
Comment (zdravko-georgiev) [inline packages/preview-middleware-client/src/cpe/changes/validator.ts:31]: If I enter `{abc}` then I do not get an error. Is `{abc}` valid?

Comment (GDamyanov) [reply packages/preview-middleware-client/src/cpe/changes/validator.ts:31]: Yes, because there could be bindings for other models different than i18n

**Review by zdravko-georgiev — APPROVED** [review]
Code is well structured and easy to understand
Very good test coverage
Manual testing ✅

**Pull request merged**